File: agent/agent_service.py
import logging
from typing import Dict, Any
from agent.nlu_parser import parse_intent_llm
from gate.models import AuthorizeRequest, GateDecision
from gate.gate_service import authorize_transaction, format_inr

logger = logging.getLogger(__name__)

def process_agent_request(user_prompt: str, user_id_override: str = None) -> Dict[str, Any]:
    """
    Agent workflow:
    1. Parse natural language intent into structured parameters.
    2. Forward request ONLY to the Mandate Spend-Gate service (never to Razorpay directly).
    3. Return plain language explanation of the gate's decision.
    """
    intent = parse_intent_llm(user_prompt)
    
    user_id = user_id_override if user_id_override else intent["user_id"]
    merchant_id = intent["merchant_id"]
    amount = intent["amount"]

    if amount <= 0:
        return {
            "status": "error",
            "prompt": user_prompt,
            "agent_message": "I couldn't identify a valid transaction amount from your request. Please specify an amount, e.g., 'Spend ₹500 at Zepto'.",
            "gate_decision": None,
            "parsed_intent": intent
        }

    # Prepare gate authorization request
    auth_req = AuthorizeRequest(
        user_id=user_id,
        merchant_id=merchant_id,
        amount=amount
    )

    # Call Gate Service
    decision: GateDecision = authorize_transaction(auth_req)
    dec_dict = decision.to_dict()

    # Build human explanation
    if decision.decision == "approved":
        agent_msg = (
            f"✅ **Transaction Approved by Spend-Gate**\n"
            f"- Merchant: `{merchant_id.upper()}`\n"
            f"- Amount: `{format_inr(amount)}`\n"
            f"- Mandate ID: `{decision.mandate_id}`\n"
            f"- Razorpay Order ID: `{decision.razorpay_order_id}`\n\n"
            f"The payment authorization passed all 5 mandate checks and the Razorpay test order has been generated."
        )
    elif decision.decision == "denied":
        agent_msg = (
            f"🚫 **Transaction Blocked by Spend-Gate**\n"
            f"- Reason: `{decision.reason}`\n"
            f"- Merchant: `{merchant_id.upper()}`\n"
            f"- Requested Amount: `{format_inr(amount)}`\n"
            f"- Mandate ID: `{decision.mandate_id}`\n\n"
            f"**Detail:** {decision.message}"
        )
    else:
        agent_msg = (
            f"⚠️ **Gate Service Error**\n"
            f"- Reason: `{decision.reason}`\n"
            f"- Detail: {decision.message}"
        )

    return {
        "status": "completed",
        "prompt": user_prompt,
        "agent_message": agent_msg,
        "parsed_intent": {
            "user_id": user_id,
            "merchant_id": merchant_id,
            "amount_paise": amount,
            "amount_formatted": format_inr(amount),
            "parser_used": intent.get("parser", "unknown")
        },
        "gate_decision": dec_dict
    }

File: src/agent_service.rs
use crate::config::Config;
use crate::database::DbConnection;
use crate::models::{AuthorizeRequest, GateDecision};
use crate::nlu_parser::parse_intent_fallback;
use crate::gate_service::{authorize_transaction, format_inr};

pub fn process_agent_request(prompt: &str, user_id_override: Option<&str>, cfg: &Config) -> (String, GateDecision) {
    let mut intent = parse_intent_fallback(prompt);
    if let Some(uid) = user_id_override {
        intent.user_id = uid.to_string();
    }

    let conn = DbConnection::open(&cfg.db_path).unwrap();

    if intent.amount_paise <= 0 {
        let dec = GateDecision {
            decision: "error".to_string(),
            reason: "invalid_amount".to_string(),
            mandate_id: None,
            user_id: intent.user_id.clone(),
            merchant_id: intent.merchant_id.clone(),
            requested_amount: 0,
            limit: None,
            already_spent: None,
            razorpay_order_id: None,
            updated_mandate: None,
            message: Some("Invalid transaction amount in prompt.".to_string()),
        };
        let msg = "I couldn't identify a valid transaction amount from your request. Please specify an amount, e.g., 'Spend ₹500 at Zepto'.".to_string();
        return (msg, dec);
    }

    let req = AuthorizeRequest {
        user_id: intent.user_id.clone(),
        merchant_id: intent.merchant_id.clone(),
        amount: intent.amount_paise,
        mandate_id: None,
    };

    let decision = authorize_transaction(&conn, &req, cfg);

    let message = if decision.decision == "approved" {
        format!(
            "✅ **Transaction Approved by Spend-Gate**\n- Merchant: `{}`\n- Amount: `{}`\n- Mandate ID: `{}`\n- Razorpay Order ID: `{}`\n\nThe payment authorization passed all 5 mandate checks and the Razorpay test order has been generated.",
            intent.merchant_id.to_uppercase(),
            format_inr(intent.amount_paise),
            decision.mandate_id.as_deref().unwrap_or("none"),
            decision.razorpay_order_id.as_deref().unwrap_or("none")
        )
    } else if decision.decision == "denied" {
        format!(
            "🚫 **Transaction Blocked by Spend-Gate**\n- Reason: `{}`\n- Merchant: `{}`\n- Requested Amount: `{}`\n- Mandate ID: `{}`\n\n**Detail:** {}",
            decision.reason,
            intent.merchant_id.to_uppercase(),
            format_inr(intent.amount_paise),
            decision.mandate_id.as_deref().unwrap_or("none"),
            decision.message.as_deref().unwrap_or("Denied")
        )
    } else {
        format!(
            "⚠️ **Gate Service Error**\n- Reason: `{}`\n- Detail: {}",
            decision.reason,
            decision.message.as_deref().unwrap_or("Error")
        )
    };

    (message, decision)
}
